Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,91 @@
 
 #include "config.h"
 #include "ScrollAnchoringController.h"
+#include "ElementIterator.h"
+#include "HTMLHtmlElement.h"
+#include "RenderBox.h"
+#include "TypedElementDescendantIteratorInlines.h"
+#include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
 void ScrollAnchoringController::invalidateAnchorElement()
 {
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    if (!m_midUpdatingScrollPositionForAnchorElement) {
+        m_anchorElement = nullptr;
+        m_lastOffsetForAnchorElement = { };
+        m_isQueuedForScrollPositionUpdate = false;
+        m_frameView.queueScrollableAreaForScrollAnchoringUpdate(m_frameView);
+    }
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
 {
-    // TODO: implement
-    // perform Anchor Element Selection algorithm on dom elements within frame view's viewable region
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return ScrollOffset(FloatPoint(candidate.absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location()));
+}
+
+void ScrollAnchoringController::chooseAnchorElement()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+
+    auto* document = m_frameView.frame().document();
+    if (!document)
+        return;
+
+    Element* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
+        if (auto renderer = element.renderBox()) {
+            if (renderer->style().overflowAnchor() == OverflowAnchor::None || &element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
+                continue;
+            auto boxRect = renderer->absoluteBoundingBoxRect();
+            if (layoutViewport.intersects(boxRect))
+                candidateAnchor = &element;
+            if (layoutViewport.contains(boxRect))
+                break;
+        }
+    }
+    if (!candidateAnchor)
+        return;
+    m_anchorElement = candidateAnchor;
+    m_lastOffsetForAnchorElement = computeOffset(*m_anchorElement->renderBox());
+}
+
+void ScrollAnchoringController::updateAnchorElement()
+{
+    if (frameView().scrollPosition().isZero() || m_isQueuedForScrollPositionUpdate)
```

Comment:
This will be wrong when there's a ScrollAnchoringController per overflow area

Reply:
Same as above.